### ISSUE TITLE
IDE0051 false positive in recursive pattern matching

### ISSUE THREAD

**Issue by genlu** [dotnet/roslyn#32981]
Both fields in the code sample below are flagged as unused private members. Note that FAR correctly reports them as "read" references in recursive pattern matching.

```cs
    public class Foo
    {
        string field1;
        string field2;

        public string RecPattern(Foo a)
        {
            if (a is Foo { field2: null, field1: string f1 })
                return f1;
            return "";
        }
    }
```

**Comment by mavasani** [comment]
> Note that FAR correctly reports them as "read" references in recursive pattern matching.

That seems to be happening because FAR reference kind column defaults to Read for any references to locals/params/fields/properties that it does not recognize as a write.

**Comment by mavasani** [comment]
Tagging @gafter @AlekseyTs @333fred @jinujoseph @KathleenDollard 

This bug manifests in the IDE analyzer due to the fact the operation sub-tree for an `IRecursivePatternOperation` has no `IFieldReferenceOperation` for references to `field1` or `field2` in `a is Foo { field2: null, field1: string f1 }`.

Based on our last IOperation design meeting about API shape for property sub-patterns, we decided the type of [IRecursivePatternOperation.PropertySubpatterns](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IRecursivePatternOperation.cs,36) will be changed to an array of operations, potentially `ImmutableArray<IPropertySubPatternOperation> PropertySubPatterns { get; }`. Is that correct?

However, I can't recall what was the final shape of `IPropertySubPatternOperation` that we settled on:
1. `IPropertySubPatternOperation` will expose a child `IMemberReferenceOperation` for the reference to field or property OR 
2. `IPropertySubPatternOperation` will derive from `IMemberReferenceOperation` and expose the symbol and instance operation (with a special instance reference kind).

If it is the former shape, then this bug will automatically get fixed when that change goes in as we will see an `IFieldReferenceOperation` in the operation tree such that it is not target of an assignment operation, (hence classified as a read operation by this analyzer).
If it is the latter, then the analyzer needs to be updated to understand `IPropertySubPatternOperation` and then look for symbol on it.

I kind of like the former API shape as all existing analyzers that register for an `OperationKind.FieldReference` or `OperationKind.PropertyReference` will automatically get callbacks for references to field/property within recursive patterns, which seems beneficial to me.

**Comment by 333fred** [comment]
@mavasani we decided on shape 1, it will expose an `IMemberReferenceOperation` child when one exists.

**Comment by mavasani** [comment]
Thanks @333fred. @gafter Can you notify me once the new recursive pattern IOperation API is checked in? I will add a unit test to verify it fixes this bug.

**Comment by mavasani** [comment]
@genlu I verified this no longer repros on latest 16.1 preview bits.
